### DISCUSSION TITLE
Get machines for license - C++ code suddenly not working

### DISCUSSION THREAD

**Discussion by asaf-b** [keygen-sh/community#58] [Issue]
Hi Zeke,
Today I got a few users of my product report about a weird issue they are having relating to log-in/log-out.
These reports were a surprise, as they didn't come from a version update from me, but really "out of the blue".
After I reproduced the issue and checked what went wrong- I found the problem, although I can't figure out why.

I'm using C++ to make my requests to Keygen.
The source of the issue was that my program could not delete machines for users to log them out. Once again nothing changed in my code.
When debugging, I noticed that my "list machines for license" request was no longer working. I keep getting empty responses, as well as "not an object" errors.
Here is what I do (which is still the method shown in the API documentation here: https://keygen.sh/docs/api/machines/#machines-list):
```
uri_builder uriMac(U("/machines"));
uriMac.append_query(U("limit"), 1);
uriMac.append_query(U("key"), licKey.as_string());
req.set_request_uri(uriMac.to_uri());
req.set_method(methods::GET);
```
This code is suddenly not working. Were there any recent changes on Keygen that might cause this? 
I literally didn't touch this code for at least 4 months, and there were no version releases for 3 months.
I'd appreciate your help trying to figure out what is going wrong here.

Thank you,
Asaf Ben Zur

**Comment by asaf-b** [comment]
Just a quick update- I just tested the exact same request, with the same parameters (same account, same license key) on Python, and everything worked just fine. I don't know why just the C++ code suddenly doesn't work.

**Reply by ezekg** [reply]
Probably because we fixed the issue. Try again from C++ and let me know? 😆

**Comment by ezekg** [comment]
Hey Asaf, appreciate the report. We re-encrypted license keys as part of a normal key rotation procedure and it looks like a small subset of licenses were not rotated correctly, which was causing this error when filtering machines by license key, unfortunately due to a bug in Rails (which we're working on a report for). Should be fixed now. Sorry for the trouble.

**Comment by asaf-b** [comment]
This is what I call **LIGHTNING SPEED**. 
Everything is back in working order in C++ code. 
You are the best Zeke thanks so much!

**Reply by ezekg** [reply]
No problem. Glad it's fixed! Again, sorry for the trouble. We wrote a few tests to make sure this doesn't happen again.

**Comment by JackBowis** [comment]
I think this may have happened again.

I'm working on a C++ app, and the "List All Machines" command seems to fail (returns a JSON response that just says "data"), when all other commands seem to be working fine.

After seeing this post, I sent a CURL command on my Windows computer:

_curl -H "Content-Type: application/vnd.api+json" -H "Accept: application/vnd.api+json" -H "Authorization: License |licenseKey|" https://api.keygen.sh/v1/accounts/|accountID|/machines?key=|licenseKey|](https://api.keygen.sh/v1/accounts/|accountID|/machines?key=|licenseKey|) -g_

Which was successful, and received a response with the machine associated with the license.

Can you guys check if something happened to make this command fail from C++ code again?

**Reply by ezekg** [reply]
@JackBowis this seems like a separate issue. Your request is [listing machines](https://keygen.sh/docs/api/machines/#machines-list) for a given license key using a `key` query parameter. An empty response would indicate there are no matching machines for that key. This could be an issue with the encoding of the license key if it contains special non-URL-friendly characters, or the license simply has no machines associated with it. 

Either way, this doesn't indicate an issue on our end. Feel free to share the request ID and I can help further, or we can move this to Discord or email, or a separate discussion since your issue seems unrelated to this one.

**Reply by JackBowis** [reply]
I think I found my problem. I wasn't accounting for the fact that the response for "List All Machines" gave an array, which changes the way my code needs to convert it into JSON format and parse it for the desired data. My recursive printing of the JSON response objects failed to print anything in the array (since it's a JSON array, not a JSON object), and I didn't immediately notice that the raw printout of the CURL command included the square braces. Once I just printed out the unformatted response in my code, I realized my problem.

Thanks for the response, and hopefully anyone else having trouble with the Get Machines For License query can learn from my error: it returns an array, so you have to parse it differently from all the other responses!
